app/scripts/ui.js

```diff
@@ -61,11 +56,107 @@ async function start () {
     }
   }
 
-  function displayCriticalError (err) {
+  function displayCriticalError (container, err) {
     container.innerHTML = '<div class="critical-error">The MetaMask app failed to load: please open and close MetaMask again to restart.</div>'
     container.style.height = '80px'
     log.error(err.stack)
     throw err
   }
 
+  function initializeUiWithTab (tab) {
+    const container = document.getElementById('app-content')
+    initializeUi(tab, container, connectionStream, (err, store) => {
+      if (err) {
+        return displayCriticalError(container, err)
+      }
+
+      const state = store.getState()
+      const { metamask: { completedOnboarding } = {} } = state
+
+      if (!completedOnboarding && windowType !== ENVIRONMENT_TYPE_FULLSCREEN) {
+        global.platform.openExtensionInBrowser()
+      }
+    })
+  }
+}
+
+async function queryCurrentActiveTab (windowType) {
+  return new Promise((resolve) => {
+    // At the time of writing we only have the `activeTab` permission which means
+    // that this query will only succeed in the popup context (i.e. after a "browserAction")
+    if (windowType !== ENVIRONMENT_TYPE_POPUP) {
+      resolve({})
+      return
+    }
+
+    extension.tabs.query({active: true, currentWindow: true}, (tabs) => {
+      const [activeTab] = tabs
+      const {title, url} = activeTab
+      const origin = url ? urlUtil.parse(url).hostname : null
+      resolve({
+        title, origin, url,
+      })
+    })
+  })
+}
+
+function initializeUi (activeTab, container, connectionStream, cb) {
+  connectToAccountManager(connectionStream, (err, backgroundConnection) => {
+    if (err) {
+      return cb(err)
+    }
+
+    launchMetaMaskUi({
+      activeTab,
+      container,
+      backgroundConnection,
+    }, cb)
+  })
+}
+
+/**
+ * Establishes streamed connections to background scripts and a Web3 provider
```

Comment:
Everything below here is just moved from `popup-core.js` correct?

Reply:
In the separate notification bundle PR, I did essentially the opposite of this: I moved more things from `ui.js` into `popup-core.js`, because they were shared between `ui.js` and the newly created `notification.js`.

I'm guessing they were merged here because the separation wasn't useful? It will shortly become useful, so it might be worth moving everything back to `popup-core.js`. Oryou can not bother, and I'll just do that when resolving this conflict in my PR - I don't mind.

Reply:
> Everything below here is just moved from `popup-core.js` correct?

Yes, that file was only used here so I merged it/inlined it. The name wasn't quite accurate and it only had this single use.

I think this diff and your diff re-introducing separate files might be good together (that is, I'll leave this for now). It seems, er, semantically correct to me that we'd split things out later if needed.